WangscapeTest/TestCartesianPower.cpp

```diff
@@ -0,0 +1,262 @@
+    EXPECT_NE(cp.cbegin(), it);
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonOther)
+{
+    tilegen::CartesianPower<Iterator> cp1{base.cbegin() + 1, 2, 4};
+    tilegen::CartesianPower<Iterator> cp2{base.cbegin(), base.size(), 3};
+    EXPECT_NE(cp.begin(), cp1.begin());
+    EXPECT_NE(cp.begin(), cp2.begin());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerComparisonOtherMid)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 23);
+    EXPECT_EQ(it, it);
+    EXPECT_NE(it, cp.cbegin());
+    EXPECT_NE(cp.cbegin(), it);
+    EXPECT_NE(it, cp.cend());
+    EXPECT_NE(cp.cend(), it);
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerRange)
+{
+    auto first = base.cbegin();
+    auto size = base.size();
+    EXPECT_EQ(first, cp.first);
+    EXPECT_EQ(size, cp.baseSize);
+    EXPECT_EQ(first, cp.cbegin().getFirst());
+    EXPECT_EQ(size, cp.cbegin().getBaseSize());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValueBegin)
+{
+    auto it = cp.cbegin();
+    EXPECT_EQ(4, it.getValue(0));
+    EXPECT_EQ(4, it.getValue(1));
+    EXPECT_EQ(4, it.getValue(2));
+    EXPECT_EQ(4, it.getValue(3));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValueBegin1)
+{
+    auto it = cp.cbegin();
+    ++it;
+    EXPECT_EQ(5, it.getValue(0));
+    EXPECT_EQ(4, it.getValue(1));
+    EXPECT_EQ(4, it.getValue(2));
+    EXPECT_EQ(4, it.getValue(3));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValueBegin01)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 3);
+    EXPECT_EQ(4, it.getValue(0));
+    EXPECT_EQ(5, it.getValue(1));
+    EXPECT_EQ(4, it.getValue(2));
+    EXPECT_EQ(4, it.getValue(3));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValueLast)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 80);
+    EXPECT_EQ(6, it.getValue(0));
+    EXPECT_EQ(6, it.getValue(1));
+    EXPECT_EQ(6, it.getValue(2));
+    EXPECT_EQ(6, it.getValue(3));
+    ++it;
+    EXPECT_EQ(it, cp.cend());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerCoordinates)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 15);
+
+    EXPECT_EQ(0, it.getCoordinate(0));
+    EXPECT_EQ(2, it.getCoordinate(1));
+    EXPECT_EQ(1, it.getCoordinate(2));
+    EXPECT_EQ(0, it.getCoordinate(3));
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerArrow)
+{
+    auto it = cp.cbegin();
+    EXPECT_EQ(cp.power, it->size());
+}
+
+TEST_F(TestCartesianPower, TestCartesianPowerValues)
+{
+    auto it = cp.cbegin();
+    std::advance(it, 15);
+    const std::vector<int>& values = *it;
+
+    EXPECT_EQ(4, values[0]);
+    EXPECT_EQ(6, values[1]);
+    EXPECT_EQ(5, values[2]);
+    EXPECT_EQ(4, values[3]);
+
+    for (size_t i = 0; i < values.size(); i++)
+    {
+        EXPECT_EQ(values[i], it.getValue(i));
+    }
+
+    for (size_t i = 0; i < values.size(); i++)
+    {
+        EXPECT_EQ(values[i], *it.getIterator(i));
+    }
+
+    for (size_t i = 0; i < values.size(); i++)
+    {
+        EXPECT_EQ(values[i], *it.getIterators()[i]);
```

Comment:
I think as long as `getValue`, `getValues`, `getIterator`, and `getIterators` are in `CartesianPowerIterator`'s public API, there should be tests to ensure that they're consistent.

I think it's possible to make `getIterator[s]` `private`, though.